modules/ui/field.js

```diff
@@ -171,10 +173,15 @@ export function uiField(context, presetField, entityIDs, options) {
             }
         }
 
+        var sourceSubfield = uiSourceSubfield(field, _tags, dispatch);
 
         // Update
         container = container
-            .merge(enter);
+        .merge(enter);
+        
+        if(options.source){
+            sourceSubfield.button(labelEnter, container);
```

Comment:
Placing this button in the field’s header row makes some sense, since it’s a standard feature of any field. After all, if we were to place it in the main row of the field, we’d have to contend with the many data types that result in different field layouts. But since the header row isn’t next to the location where the source subfield appears, we’ll need to visually connect the two elements somehow.

Feel free to explore a unique look and feel for this subfield if you think that’ll help. You can describe it or create a mockup to get some feedback on the idea.

Reply:
I could try moving the button underneath the main row of the field or maybe turn the source field into a row between the field label and the main row of the field.
Another possibility is to remove the source button after it is clicked and the source subfield appears and return it if the subfield is deleted. This way, we wouldn't have to make the button and the field be visually related, because only one of them would be visible at a time.
Another possibility is to add the icon that is inside the source button somewhere in the label of the source subfield. This way, they would be apart, but having the same icon would make them visually connected.
What do you think of these possibilities?

Reply:
These ideas are all worth exploring if you have time. As you try out different layouts for this feature, consider the different states that a user might see even without interacting with a source. For example, the user might select a feature that already has sources for some fields, so they need to know how the source relates to the rest of the field even though they weren’t the one who put it there in the first place. If they happen to see these filled-out sources on one feature, then select a feature that has no sources, will they intuitively know how to add a source?

On the forum thread, Jeff likened the source subfields to Wikidata’s “‘add reference’ capability”. [Wikidata](https://www.wikidata.org/) is a pretty important project for linked data. A number of OSM and OHM mappers are also contributors to Wikidata. I think Jeff is referring to how Wikidata lets you attach [sources](https://www.wikidata.org/wiki/Help:Sources) to individual factual statements. Unfortunately, Wikidata’s UI doesn’t use an icon to represent sources or references, but maybe the way it lays out existing references would be of interest to you.

For example, [this item](https://www.wikidata.org/wiki/Q25384440?uselang=en#P2561) has two name statements with references in a collapsable section below each statement. The text inside the references section is also subtly smaller than the text outside it.

<img src="https://github.com/OpenHistoricalMap/iD/assets/1231218/2d871156-84ea-43a3-8f5c-9c3bd903a011" width="910" alt="Name is set to two different values, OpenHistoricalMap and Open Historical Map. Under the first, there is one reference, collapsed. Under the second, two references, expanded.">

Even when collapsed, it takes up a lot more space than the button you’ve implemented. After all, this UI occupies the whole page, not just a sidebar. But there might be a way to at least hide the source subfields from the preview that appears when hovering over a feature, when we need the sidebar contents to be as compact as possible.

This will be tricky to get right, but let’s arrive at a design that’s usable enough. Then once you redesign the main Source field, we’ll have a better idea of how the more basic source subfields should behave.